File: php/bin/agreement_check.php
<?php
// Reads the shared hybrid key-agreement vector
// (spec/corpus/transcripts/handshake-agreement.json) and confirms this PHP
// implementation reproduces the transcript checkpoints and transport keys.
// Sequence: mixHash(mesh, ei_pub, ki_ek, n); mixHash(er_pub); mixKey(ss_dh);
// mixHash(ct); mixKey(ss_kem); split(). Invoked by interop/check-agreement-php.sh.
require __DIR__ . '/../src/autoload.php';

use Bonemesh\KeySchedule;

$path = $argv[1] ?? null;
if ($path === null) {
    fwrite(STDERR, "usage: agreement_check <handshake-agreement.json>\n");
    exit(2);
}
$doc = json_decode(file_get_contents($path), true);
$in = fn (string $k) => hex2bin($doc['inputs'][$k]);
$out = $doc['outputs'];
$failures = 0;
$check = function (string $name, string $want, string $got) use (&$failures) {
    if (bin2hex($got) === $want) {
        echo "PASS $name\n";
    } else {
        echo "FAIL $name\n  got:  " . bin2hex($got) . "\n  want: $want\n";
        $failures++;
    }
};

$s = new KeySchedule();
$s->mixHash($in('mesh_hex'));
$s->mixHash($in('ei_pub_hex'));
$s->mixHash($in('ki_ek_hex'));
$s->mixHash($in('n_hex'));
$check('h_after_msg1', $out['h_after_msg1'], $s->h);

$s->mixHash($in('er_pub_hex'));
$s->mixKey($in('ss_dh_hex'));
$check('ck_after_dh', $out['ck_after_dh'], $s->ck);

$s->mixHash($in('kem_ct_hex'));
$s->mixKey($in('ss_kem_hex'));
$check('ck_after_kem', $out['ck_after_kem'], $s->ck);

// The msg-2 checkpoint is taken once both message-2 hash inputs (responder
// ephemeral and KEM ciphertext) are absorbed; mixKey does not alter h.
$check('h_after_msg2_ephemerals', $out['h_after_msg2_ephemerals'], $s->h);

[$i2r, $r2i] = $s->split();
$check('transport_key_i2r', $out['transport_key_i2r'], $i2r);
$check('transport_key_r2i', $out['transport_key_r2i'], $r2i);

if ($failures > 0) {
    fwrite(STDERR, "$failures checkpoint(s) mismatched\n");
    exit(1);
}
echo "hybrid key agreement reproduces every shared checkpoint\n";

File: python/bin/agreement_check.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.12"
# dependencies = ["bonemesh"]
#
# [tool.uv.sources]
# bonemesh = { path = "..", editable = true }
# ///
"""Reads the shared hybrid key-agreement vector
(spec/corpus/transcripts/handshake-agreement.json) and confirms this
implementation reproduces the transcript checkpoints and transport keys.

Sequence (security.md §5): mix_hash(mesh, ei_pub, ki_ek, n); mix_hash(er_pub);
mix_key(ss_dh); mix_hash(kem_ct); mix_key(ss_kem); split().

The X25519 secret is DERIVED from ei_priv and er_pub rather than read out of the
vector, then compared against the vector's ss_dh_hex. That is a second oracle the
schedule alone cannot give: feeding the vector's own ss_dh straight into mix_key
would still pass on an implementation whose X25519 agreement was broken, because
nothing would ever have computed it. Go, Java and Elixir derive it the same way.

Invoked by interop/check-agreement-python.sh.
"""

import json
import sys

from cryptography.hazmat.primitives.asymmetric import x25519

from bonemesh.crypto import x25519_agree
from bonemesh.keyschedule import KeySchedule


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: agreement_check <handshake-agreement.json>", file=sys.stderr)
        return 2
    with open(sys.argv[1], encoding="utf-8") as fh:
        doc = json.load(fh)
    i, o = doc["inputs"], doc["outputs"]
    hx = bytes.fromhex
    failures = 0

    def check(name: str, got: str, want: str) -> None:
        nonlocal failures
        if got == want:
            print(f"PASS {name}")
        else:
            print(f"FAIL {name}\n  got:  {got}\n  want: {want}")
            failures += 1

    priv = x25519.X25519PrivateKey.from_private_bytes(hx(i["ei_priv_hex"]))
    ss_dh = x25519_agree(priv, hx(i["er_pub_hex"]))
    check("ss_dh (derived)", ss_dh.hex(), i["ss_dh_hex"])

    s = KeySchedule()
    s.mix_hash(hx(i["mesh_hex"]))
    s.mix_hash(hx(i["ei_pub_hex"]))
    s.mix_hash(hx(i["ki_ek_hex"]))
    s.mix_hash(hx(i["n_hex"]))
    check("h_after_msg1", s.h.hex(), o["h_after_msg1"])

    s.mix_hash(hx(i["er_pub_hex"]))
    s.mix_key(ss_dh)
    check("ck_after_dh", s.ck.hex(), o["ck_after_dh"])

    s.mix_hash(hx(i["kem_ct_hex"]))
    s.mix_key(hx(i["ss_kem_hex"]))
    check("ck_after_kem", s.ck.hex(), o["ck_after_kem"])

    # The msg-2 checkpoint is taken once both message-2 hash inputs (responder
    # ephemeral and KEM ciphertext) are absorbed; mix_key does not alter h.
    check("h_after_msg2_ephemerals", s.h.hex(), o["h_after_msg2_ephemerals"])

    i2r, r2i = s.split()
    check("transport_key_i2r", i2r.hex(), o["transport_key_i2r"])
    check("transport_key_r2i", r2i.hex(), o["transport_key_r2i"])

    if failures:
        print(f"{failures} checkpoint(s) mismatched", file=sys.stderr)
        return 1
    print("hybrid key agreement reproduces every shared checkpoint")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
